### PR TITLE
instancetype: Introduce `PreferredEfi` to API

### LINKED ISSUE #12743: Unable to configure EFI persistence from preferences
/area instancetype

This is being reported as a bug given the impact it is having on Windows  11 24H2 VMs using bitlocker encryption and the `windows.11` family of common-instancetype preferences. While a workaround does exist in the form of users manually configuring `EFI` from within their `VirtualMachine` it would be cleaner to land and backport the missing support in preferences in the longer term.

**What happened**:

At present prefernces have no way of configuring EFI persistence. The only available fields being `PreferredUseEFi` and `PreferredUseSecureBoot`, both poorly named bools. 

**What you expected to happen**:

Preferences should be able to configure EFI persistence and any future EFI configurables as they appear in the `VirtualMachine` API.

**How to reproduce it (as minimally and precisely as possible)**:

* Create a Windows 11 24H2 `VirtualMachine` using both bitlocker encryption and the `windows.11` preference
*  Reboot the `VirtualMachine`
* `VirtualMachine` fails to reboot

**Additional context**:

Downstream private tracker: https://issues.redhat.com/browse/CNV-44264

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lyarwood**
### What this PR does

Before this PR:

The introduction of PreferredUseEfi and PreferredUseSecureBoot with the
original flavors API was a mistake. PreferredUseEfi in particular was
shortsighted and restrictive, being unable to express any newly
introduced attributes of the EFI struct within the core API.

After this PR:

This change introduces PreferredEfi as an eventual replacement for these
fields allowing us to express all EFI related attributes easily within a
preference.

Fixes #12743

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer

At present I do plan on backporting this to at least the `release-1.3` branch after reading over our guidelines below:

https://github.com/kubevirt/kubevirt/blob/main/docs/release-branch-backporting.md#bug-fix-definition

IMHO this is a additive and thus backwardly compatible API change that is fixing a serious shortcoming in the preferences API.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
A new `PreferredEfi` field has been added to preferences to express the preferred `EFI` configuration for a given `VirtualMachine`.
```

### CODE REVIEW

File: pkg/instancetype/instancetype.go [hunk 1]
```diff
@@ -1534,28 +1534,36 @@ func applyFirmwarePreferences(preferenceSpec *instancetypev1beta1.VirtualMachine
...
+
 	if vmiSpec.Domain.Firmware == nil {
 		vmiSpec.Domain.Firmware = &virtv1.Firmware{}
 	}
 
+	vmiFirmware := vmiSpec.Domain.Firmware
+
 	if vmiSpec.Domain.Firmware.Bootloader == nil {
-		vmiSpec.Domain.Firmware.Bootloader = &virtv1.Bootloader{}
+		vmiFirmware.Bootloader = &virtv1.Bootloader{}
```
Comment (0xFelix) [inline pkg/instancetype/instancetype.go:1546]: Use `vmiFirmare` in line above as well?

File: pkg/instancetype/instancetype.go [hunk 2]
```diff
@@ -1534,28 +1534,36 @@ func applyFirmwarePreferences(preferenceSpec *instancetypev1beta1.VirtualMachine
...
+	}
+
+	if firmware.PreferredUseBios != nil && *firmware.PreferredUseBios && vmiFirmware.Bootloader.BIOS == nil && vmiFirmware.Bootloader.EFI == nil {
+		vmiFirmware.Bootloader.BIOS = &virtv1.BIOS{}
 	}
 
-	if preferenceSpec.Firmware.PreferredUseBios != nil && *preferenceSpec.Firmware.PreferredUseBios && vmiSpec.Domain.Firmware.Bootloader.BIOS == nil && vmiSpec.Domain.Firmware.Bootloader.EFI == nil {
-		vmiSpec.Domain.Firmware.Bootloader.BIOS = &virtv1.BIOS{}
+	if firmware.PreferredUseBiosSerial != nil && vmiFirmware.Bootloader.BIOS != nil && vmiFirmware.Bootloader.BIOS.UseSerial == nil {
+		vmiFirmware.Bootloader.BIOS.UseSerial = ptr.To(*firmware.PreferredUseBiosSerial)
```
Comment (0xFelix) [inline pkg/instancetype/instancetype.go:1554]: Use kubevirt pointer package?

File: pkg/instancetype/instancetype.go [hunk 3]
```diff
@@ -1534,28 +1534,36 @@ func applyFirmwarePreferences(preferenceSpec *instancetypev1beta1.VirtualMachine
...
 
-	if preferenceSpec.Firmware.PreferredUseBiosSerial != nil && vmiSpec.Domain.Firmware.Bootloader.BIOS != nil && vmiSpec.Domain.Firmware.Bootloader.BIOS.UseSerial == nil {
-		vmiSpec.Domain.Firmware.Bootloader.BIOS.UseSerial = ptr.To(*preferenceSpec.Firmware.PreferredUseBiosSerial)
+	if firmware.PreferredEfi != nil {
+		vmiFirmware.Bootloader.EFI = firmware.PreferredEfi.DeepCopy()
 	}
 
-	if preferenceSpec.Firmware.PreferredUseEfi != nil && *preferenceSpec.Firmware.PreferredUseEfi && vmiSpec.Domain.Firmware.Bootloader.EFI == nil && vmiSpec.Domain.Firmware.Bootloader.BIOS == nil {
-		vmiSpec.Domain.Firmware.Bootloader.EFI = &virtv1.EFI{}
+	if firmware.DeprecatedPreferredUseEfi != nil && *firmware.DeprecatedPreferredUseEfi && vmiFirmware.Bootloader.EFI == nil && vmiFirmware.Bootloader.BIOS == nil {
```
Comment (0xFelix) [inline pkg/instancetype/instancetype.go:1561]: Should we make the old and new fields mutually exclusive? Or output a warning in the admitter?

File: pkg/instancetype/instancetype.go (hunk 2 above)
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:1554]: Yup sorry rebased on https://github.com/kubevirt/kubevirt/pull/12740

File: pkg/instancetype/instancetype.go (hunk 3 above)
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:1561]: I think the current behaviour is enough, ignoring the use of the deprecated fields if the newer field is used by checking if `vmiFirmware.Bootloader.EFI == nil` here.

Comment (lyarwood) [reply pkg/instancetype/instancetype.go:1561]: Ah wait, PreferredUseSecureBoot isn't using that conditional. Let me rework this slightly.

Comment (lyarwood) [reply pkg/instancetype/instancetype.go:1563]: Added the following to just return early when using `PreferredEfi`:

https://github.com/kubevirt/kubevirt/pull/12739/commits/3a31137abe464d1560e9b693e91d311273ec69cd#diff-5e2591267ff7e9303f97ccac213986c2157131e3487a04a122b7662d401f7a57R1559

**Comment by lyarwood** [comment]
/hold

I might break this up to allow the first two commits to be backported.

File: pkg/instancetype/instancetype.go [hunk 4]
```diff
@@ -1534,28 +1534,38 @@ func applyFirmwarePreferences(preferenceSpec *instancetypev1beta1.VirtualMachine
...
 
+	firmware := preferenceSpec.Firmware
+
 	if vmiSpec.Domain.Firmware == nil {
 		vmiSpec.Domain.Firmware = &virtv1.Firmware{}
 	}
 
-	if vmiSpec.Domain.Firmware.Bootloader == nil {
-		vmiSpec.Domain.Firmware.Bootloader = &virtv1.Bootloader{}
+	vmiFirmware := vmiSpec.Domain.Firmware
```
Comment (0xFelix) [inline pkg/instancetype/instancetype.go:1543]: Move to L1538 and use it in the block above too? Sorry, did not notice this earlier.

Comment (0xFelix) [reply pkg/instancetype/instancetype.go:1543]: Sorry for the noise, won't work if `vmiSpec.Domain.Firmware` is `nil`.

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Pull request merged**